In [1]:
import pandas as pd
from ucimlrepo import fetch_ucirepo 

## Fetching data from UCI Machine Learning repository

In [81]:
heart_disease = fetch_ucirepo(id=45)
X = heart_disease.data.features  
y = heart_disease.data.targets   

## Converting the data to pandas dataframe

In [106]:
heart_disease_df = X.copy() 
heart_disease_df['num'] = y 

# DATA EXPLORATION

In [24]:
# VIEW FIRST FEW ROWS OF A DATA
heart_disease_df.head()

age  sex  cp  trestbps  chol  fbs  restecg  thalach  exang  oldpeak  slope  \
0   63    1   1       145   233    1        2      150      0      2.3      3   
1   67    1   4       160   286    0        2      108      1      1.5      2   
2   67    1   4       120   229    0        2      129      1      2.6      2   
3   37    1   3       130   250    0        0      187      0      3.5      3   
4   41    0   2       130   204    0        2      172      0      1.4      1   

    ca  thal  num  
0  0.0   6.0    0  
1  3.0   3.0    2  
2  2.0   7.0    1  
3  0.0   3.0    0  
4  0.0   3.0    0

In [25]:
# VIEW LAST FEW ROWS OF A DATA
heart_disease_df.tail()

age  sex  cp  trestbps  chol  fbs  restecg  thalach  exang  oldpeak  \
298   45    1   1       110   264    0        0      132      0      1.2   
299   68    1   4       144   193    1        0      141      0      3.4   
300   57    1   4       130   131    0        0      115      1      1.2   
301   57    0   2       130   236    0        2      174      0      0.0   
302   38    1   3       138   175    0        0      173      0      0.0   

     slope   ca  thal  num  
298      2  0.0   7.0    1  
299      2  2.0   7.0    2  
300      2  1.0   7.0    3  
301      2  1.0   3.0    1  
302      1  NaN   3.0    0

In [26]:
# FETCHING THE INFO OF OUR DATA
heart_disease_df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 303 entries, 0 to 302
Data columns (total 14 columns):
 #   Column    Non-Null Count  Dtype  
---  ------    --------------  -----  
 0   age       303 non-null    int64  
 1   sex       303 non-null    int64  
 2   cp        303 non-null    int64  
 3   trestbps  303 non-null    int64  
 4   chol      303 non-null    int64  
 5   fbs       303 non-null    int64  
 6   restecg   303 non-null    int64  
 7   thalach   303 non-null    int64  
 8   exang     303 non-null    int64  
 9   oldpeak   303 non-null    float64
 10  slope     303 non-null    int64  
 11  ca        299 non-null    float64
 12  thal      301 non-null    float64
 13  num       303 non-null    int64  
dtypes: float64(3), int64(11)
memory usage: 33.3 KB


In [34]:
# STATISTICAL DESCRIPTION OF DATA

# Now for each column I will get a statistical analysis like count,mean etc for each column.
heart_disease_df.describe()

age         sex          cp    trestbps        chol         fbs  \
count  303.000000  303.000000  303.000000  303.000000  303.000000  303.000000   
mean    54.438944    0.679868    3.158416  131.689769  246.693069    0.148515   
std      9.038662    0.467299    0.960126   17.599748   51.776918    0.356198   
min     29.000000    0.000000    1.000000   94.000000  126.000000    0.000000   
25%     48.000000    0.000000    3.000000  120.000000  211.000000    0.000000   
50%     56.000000    1.000000    3.000000  130.000000  241.000000    0.000000   
75%     61.000000    1.000000    4.000000  140.000000  275.000000    0.000000   
max     77.000000    1.000000    4.000000  200.000000  564.000000    1.000000   

          restecg     thalach       exang     oldpeak       slope          ca  \
count  303.000000  303.000000  303.000000  303.000000  303.000000  299.000000   
mean     0.990099  149.607261    0.326733    1.039604    1.600660    0.672241   
std      0.994971   22.875003    0.469794    1.161075    0.616226    0.937438   
min      0.000000   71.000000    0.000000    0.000000    1.000000    0.000000   
25%      0.000000  133.500000    0.000000    0.000000    1.000000    0.000000   
50%      1.000000  153.000000    0.000000    0.800000    2.000000    0.000000   
75%      2.000000  166.000000    1.000000    1.600000    2.000000    1.000000   
max      2.000000  202.000000    1.000000    6.200000    3.000000    3.000000   

             thal         num  
count  301.000000  303.000000  
mean     4.734219    0.937294  
std      1.939706    1.228536  
min      3.000000    0.000000  
25%      3.000000    0.000000  
50%      3.000000    0.000000  
75%      7.000000    2.000000  
max      7.000000    4.000000

**Question**: What if we are only interested in the statistical analysis of the ages of the individuals?

In [28]:
heart_disease_df['age'].describe()

count    303.000000
mean      54.438944
std        9.038662
min       29.000000
25%       48.000000
50%       56.000000
75%       61.000000
max       77.000000
Name: age, dtype: float64

Now can say that the minimum age of an individual in the dataset is **29**.

Just like that we can extract other points as well!

## FOR EXAMPLE: 

### Age Distribution Insights

In [9]:
heart_disease_df['age'].describe()

count    303.000000
mean      54.438944
std        9.038662
min       29.000000
25%       48.000000
50%       56.000000
75%       61.000000
max       77.000000
Name: age, dtype: float64

**Mean Age**: The average age is 54 years.

**Age Range**: Ages range from 29 to 77 years, indicating a wide age span of individuals.

**Interquartile Range (IQR)**: The middle 50% of ages fall between 48 and 61 years.

### Gender Distribution

In [10]:
heart_disease_df['sex'].value_counts()

sex
1    206
0     97
Name: count, dtype: int64

##### we can see that males(1) indiviuals are in greater amount as compared to females(0) in our dataset

### Chest Pain Type (cp)

- **4 Types of chest pain in our dataset:**
  - 1: Typical angina  
  - 2: Atypical angina  
  - 3: Non-anginal pain  
  - 4: Asymptomatic

In [11]:
heart_disease_df['cp'].value_counts()

cp
4    144
3     86
2     50
1     23
Name: count, dtype: int64

It appears that, in our dataset, most individuals experienced the fourth type of chest pain, which is **asymptomatic**.

### Blood Pressure (trestbps)

In [12]:
heart_disease_df['trestbps'].describe()

count    303.000000
mean     131.689769
std       17.599748
min       94.000000
25%      120.000000
50%      130.000000
75%      140.000000
max      200.000000
Name: trestbps, dtype: float64

- **Mean Blood Pressure:**
  - Average resting blood pressure is ~131.7 mmHg.
- **Range:**
  - Blood pressure varies from 94 mmHg to 200 mmHg.

### Maximum Heart Rate Achieved (thalach)

In [13]:
heart_disease_df['thalach'].describe()

count    303.000000
mean     149.607261
std       22.875003
min       71.000000
25%      133.500000
50%      153.000000
75%      166.000000
max      202.000000
Name: thalach, dtype: float64

- **Mean:**
  - Average maximum heart rate is ~149.6 beats per minute (bpm).
- **Range:**
  - Values vary from 71 bpm to 202 bpm, indicating individuals with varying cardiac health.


### Thalassemia (thal)

In [15]:
heart_disease_df['thal'].describe()

count    301.000000
mean       4.734219
std        1.939706
min        3.000000
25%        3.000000
50%        3.000000
75%        7.000000
max        7.000000
Name: thal, dtype: float64

Mean value of thal is **4.73**, Most individuals in the dataset likely have **normal** or **less severe** heart conditions

### Target Variable (num)

- **0**: No heart disease (less than 50% narrowing)
- **1**: Mild heart disease (e.g., a very small amount of narrowing)
- **2**: Moderate heart disease (e.g., moderate narrowing)
- **3**: Severe heart disease (e.g., major narrowing)
- **4**: Very severe or extreme heart disease (e.g., total or near-total blockage)

In [32]:
heart_disease_df['num'].describe()

count    303.000000
mean       0.937294
std        1.228536
min        0.000000
25%        0.000000
50%        0.000000
75%        2.000000
max        4.000000
Name: num, dtype: float64

- Mean value of target is **0.93**, indicating most individuals likely belong to a specific class (e.g., with heart disease).

In [33]:
heart_disease_df['num'].value_counts()

num
0    164
1     55
2     36
3     35
4     13
Name: count, dtype: int64

It looks like most people in our dataset has **no heart disease**.

## Filling NAN/NA Values

In [37]:
heart_disease_df.isnull().sum()

age         0
sex         0
cp          0
trestbps    0
chol        0
fbs         0
restecg     0
thalach     0
exang       0
oldpeak     0
slope       0
ca          4
thal        2
num         0
dtype: int64

Now we can see that we have NA values in **thal** and **ca** column, Let's resolve it

In [49]:
heart_disease_df['thal'] = heart_disease_df['thal'].fillna(heart_disease_df['thal'].mean())

In [50]:
heart_disease_df.isnull().sum()

age         0
sex         0
cp          0
trestbps    0
chol        0
fbs         0
restecg     0
thalach     0
exang       0
oldpeak     0
slope       0
ca          4
thal        0
num         0
dtype: int64

In [51]:
heart_disease_df['thal'].describe()

count    303.000000
mean       4.734219
std        1.933272
min        3.000000
25%        3.000000
50%        3.000000
75%        7.000000
max        7.000000
Name: thal, dtype: float64

In [52]:
heart_disease_df['ca'].describe()

count    299.000000
mean       0.672241
std        0.937438
min        0.000000
25%        0.000000
50%        0.000000
75%        1.000000
max        3.000000
Name: ca, dtype: float64

In [53]:
heart_disease_df['ca'] = heart_disease_df['ca'].fillna(heart_disease_df['ca'].mean())

In [55]:
heart_disease_df['ca'].describe()

count    303.000000
mean       0.672241
std        0.931209
min        0.000000
25%        0.000000
50%        0.000000
75%        1.000000
max        3.000000
Name: ca, dtype: float64

In [58]:
heart_disease_df.isnull().sum()

age         0
sex         0
cp          0
trestbps    0
chol        0
fbs         0
restecg     0
thalach     0
exang       0
oldpeak     0
slope       0
ca          0
thal        0
num         0
dtype: int64

#### Question: Why did I fill the NA values with the mean of that column?

**Answer**: I filled the NA values with the mean of the column because I might later need to calculate the mean or other statistical metrics like the minimum value from that column. By filling the missing values with the mean, I ensure that these calculations are not impacted. If I had filled the NA values with a string like "Unknown," it could interfere with the functionality of the `describe` method, which performs statistical analysis on numerical data.


In [59]:
heart_disease_df.head(5)

age  sex  cp  trestbps  chol  fbs  restecg  thalach  exang  oldpeak  slope  \
0   63    1   1       145   233    1        2      150      0      2.3      3   
1   67    1   4       160   286    0        2      108      1      1.5      2   
2   67    1   4       120   229    0        2      129      1      2.6      2   
3   37    1   3       130   250    0        0      187      0      3.5      3   
4   41    0   2       130   204    0        2      172      0      1.4      1   

    ca  thal  num  
0  0.0   6.0    0  
1  3.0   3.0    2  
2  2.0   7.0    1  
3  0.0   3.0    0  
4  0.0   3.0    0

### Top 3 Women with the Highest Ages and Their Heart Disease Numbers

In [68]:
heart_disease_df[['age','sex','num']][heart_disease_df['sex']==0].sort_values(by="age",ascending=False).head(3)

age  sex  num
257   76    0    0
233   74    0    0
273   71    0    0

It looks like top 3 with highest ages have **No heart disease**.

In [72]:
heart_disease_df[heart_disease_df['sex']==0].sort_values(by="age",ascending=False).num.value_counts()

num
0    72
1     9
3     7
2     7
4     2
Name: count, dtype: int64

It looks like only 2 womens have **Very severe or extreme heart disease** in our dataset

### How many mens have very severe or extreme heart disease ?

In [73]:
heart_disease_df[heart_disease_df['sex']==1].sort_values(by="age",ascending=False).num.value_counts()

num
0    92
1    46
2    29
3    28
4    11
Name: count, dtype: int64

It looks like 11 mens have **Very severe or extreme heart disease** in our dataset

### Top 3 men with the Highest Ages and Their Heart Disease Numbers

In [75]:
heart_disease_df[['age','sex','num']][heart_disease_df['sex']==1].sort_values(by="age",ascending=False).head(3)

age  sex  num
161   77    1    4
136   70    1    4
170   70    1    3

### Converting 0 and 1 to Male and Female Strings to increase readability.

In [76]:
heart_disease_df.dtypes

age           int64
sex           int64
cp            int64
trestbps      int64
chol          int64
fbs           int64
restecg       int64
thalach       int64
exang         int64
oldpeak     float64
slope         int64
ca          float64
thal        float64
num           int64
dtype: object

In [88]:
heart_disease_df['sex'] = heart_disease_df['sex'].map({0: 'Female', 1: 'Male'})

In [90]:
heart_disease_df.head(6)

age     sex  cp  trestbps  chol  fbs  restecg  thalach  exang  oldpeak  \
0   63    Male   1       145   233    1        2      150      0      2.3   
1   67    Male   4       160   286    0        2      108      1      1.5   
2   67    Male   4       120   229    0        2      129      1      2.6   
3   37    Male   3       130   250    0        0      187      0      3.5   
4   41  Female   2       130   204    0        2      172      0      1.4   
5   56    Male   2       120   236    0        0      178      0      0.8   

   slope   ca  thal  num  
0      3  0.0   6.0    0  
1      2  3.0   3.0    2  
2      2  2.0   7.0    1  
3      3  0.0   3.0    0  
4      1  0.0   3.0    0  
5      1  0.0   3.0    0

In [92]:
heart_disease_df.dtypes

age           int64
sex          object
cp            int64
trestbps      int64
chol          int64
fbs           int64
restecg       int64
thalach       int64
exang         int64
oldpeak     float64
slope         int64
ca          float64
thal        float64
num           int64
dtype: object

We can later on convert that back to 0 and 1 when working with ML binary models!

## Analysis: Average Age and Thalassemia Status by Heart Disease Category

In [97]:
heart_disease_df[['num','age','thal']].groupby('num').mean()

age      thal
num                     
0    52.585366  3.797546
1    55.381818  5.345455
2    58.027778  6.028571
3    56.000000  6.285714
4    59.692308  6.230769